File: modules/platforms/dotnet/Apache.Ignite.Core/Configuration/SystemDataRegionConfiguration.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache.Ignite.Core.Configuration
{
    using System.ComponentModel;
    using System.Diagnostics;
    using Apache.Ignite.Core.Binary;

    /// <summary>
    /// Data storage configuration for system cache.
    /// </summary>
    public class SystemDataRegionConfiguration
    {
        /// <summary>
        /// Default size in bytes of a memory chunk reserved for system cache initially.
        /// </summary>
        public const long DefaultInitialSize = 40 * 1024 * 1024;

        /// <summary>
        /// Default max size in bytes of a memory chunk for the system cache.
        /// </summary>
        public const long DefaultMaxSize = 100 * 1024 * 1024;

        /// <summary>
        /// Initializes a new instance of the <see cref="SystemDataRegionConfiguration"/> class.
        /// </summary>
        public SystemDataRegionConfiguration()
        {
            InitialSize = DefaultInitialSize;
            MaxSize = DefaultMaxSize;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="DataStorageConfiguration"/> class.
        /// </summary>
        /// <param name="reader">The reader.</param>
        internal SystemDataRegionConfiguration(IBinaryRawReader reader)
        {
            Debug.Assert(reader != null);

            InitialSize = reader.ReadLong();
            MaxSize = reader.ReadLong();
        }
        
        /// <summary>
        /// Writes this instance to the specified writer.
        /// </summary>
        /// <param name="writer">The writer.</param>
        internal void Write(IBinaryRawWriter writer)
        {
           Debug.Assert(writer != null);
           
           writer.WriteLong(InitialSize);
           writer.WriteLong(MaxSize);
        }
        
        /// <summary>
        /// Gets or sets the size in bytes of a memory chunk reserved for system needs.
        /// </summary>
        [DefaultValue(DefaultInitialSize)]
        public long InitialSize { get; set; }
        
        /// <summary>
        /// Gets or sets the maximum memory region size in bytes reserved for system needs.
        /// </summary>
        [DefaultValue(DefaultMaxSize)]
        public long MaxSize { get; set; }
    }
}

File: modules/core/src/main/java/org/apache/ignite/configuration/SystemDataRegionConfiguration.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ignite.configuration;

import java.io.Serializable;
import org.apache.ignite.internal.util.typedef.internal.A;
import org.apache.ignite.internal.util.typedef.internal.S;

/**
 * This class allows defining system data region configuration with various parameters for Apache Ignite
 * page memory (see {@link DataStorageConfiguration}.
 * This class is similiar to {@link DataRegionConfiguration}, but with restricted set of properties.
 */
public class SystemDataRegionConfiguration implements Serializable {
    /** Serial version uid. */
    private static final long serialVersionUID = 0L;

    /** Default initial size in bytes of a memory chunk for the system cache (40 MB). */
    public static final long DFLT_SYS_REG_INIT_SIZE = 40L * 1024 * 1024;

    /** Default max size in bytes of a memory chunk for the system cache (100 MB). */
    public static final long DFLT_SYS_REG_MAX_SIZE = 100L * 1024 * 1024;

    /** Initial size in bytes of a memory chunk reserved for system cache. */
    private long initSize = DFLT_SYS_REG_INIT_SIZE;

    /** Maximum size in bytes of a memory chunk reserved for system cache. */
    private long maxSize = DFLT_SYS_REG_MAX_SIZE;

    /**
     * Initial size of a data region reserved for system cache.
     *
     * @return Size in bytes.
     */
    public long getInitialSize() {
        return initSize;
    }

    /**
     * Sets initial size of a data region reserved for system cache.
     *
     * Default value is {@link #DFLT_SYS_REG_INIT_SIZE}
     *
     * @param initSize Size in bytes.
     * @return {@code this} for chaining.
     */
    public SystemDataRegionConfiguration setInitialSize(long initSize) {
        A.ensure(initSize > 0, "System region initial size should be greater than zero.");

        this.initSize = initSize;

        return this;
    }

    /**
     * Maximum data region size in bytes reserved for system cache.
     *
     * @return Size in bytes.
     */
    public long getMaxSize() {
        return maxSize;
    }

    /**
     * Sets maximum data region size in bytes reserved for system cache. The total size should not be less than 10 MB
     * due to internal data structures overhead.
     *
     * Default value is {@link #DFLT_SYS_REG_MAX_SIZE}.
     *
     * @param maxSize Maximum size in bytes for system cache data region.
     * @return {@code this} for chaining.
     */
    public SystemDataRegionConfiguration setMaxSize(long maxSize) {
        A.ensure(maxSize > 0, "System region max size should be greater than zero.");

        this.maxSize = maxSize;

        return this;
    }

    /** {@inheritDoc} */
    @Override public String toString() {
        return S.toString(SystemDataRegionConfiguration.class, this);
    }
}
